### DISCUSSION TITLE
Is there a replacement for delta_temporality_selector, InexpensiveSelector?

### DISCUSSION THREAD

**Discussion by svix-jplatte** [open-telemetry/opentelemetry-rust#1511] [Q&A] [answered]
Hi, I'm upgrading an app from opentelemetry 0.18 to 0.21 (+ opentelemetry-otlp). Previously, it was using `InexpensiveSelector` as the aggregator selector and `ConstantTemporalitySelector` (via `delta_temporality_selector`). I can't seem to find any of those symbols anymore, were they removed? They seem easy enough to reimplement, I'm just wondering if I have to.

**Comment by cijothomas** [comment]
The terminologies are all changed for metrics (due to spec changes) and no easy way to tell what are the alternatives...
Could you describe what you are trying to achieve and we can help with that..

There are metrics-basic/advanced examples here, that showcases a lot of scenarios. It still does not show the "temporality" changing part, but that is something we will add soon. (the default temporality is Cumulative, and a lot of vendors need Delta, so probably this is what you are after?)
https://github.com/open-telemetry/opentelemetry-rust/tree/main/examples

**Reply by svix-jplatte** [reply]
Honestly I don't know what I'm looking for other than "the same behavior as before", though a colleague also pointed out that DataDog (which we're exporting to) at least _prefers_ deltas. I guess GitHub doesn't show back-references in discussions.. The change I'm talking about is [in this small PR](https://github.com/svix/svix-webhooks/pull/1184).

**Reply by cijothomas** [reply, marked as answer]
You probably need just the defaults, and pick Delta temporality given DataDog needs that.

```rust
fn init_metrics() -> metrics::Result<MeterProvider> {
    let export_config = ExportConfig {
        endpoint: ENDPOINT_URL.to_string(),
        ..ExportConfig::default()
    };
    opentelemetry_otlp::new_pipeline()
        .metrics(runtime::Tokio)
        .with_temporality_selector(DeltaTemporalitySelector::new())
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_export_config(export_config),
        )
        .with_resource(Resource::new(vec![
            KeyValue::new("service.name", SERVICE_NAME),
        ]))
        .build()
}


#[derive(Default)]
/// A temporality selector that returns Delta for all instruments except UpDownCounter and ObservableUpDownCounter.
/// This is matching OTLPExporter spec.

pub struct DeltaTemporalitySelector {}

impl DeltaTemporalitySelector {
    /// Create a new default temporality selector.
    pub fn new() -> Self {
        Self::default()
    }
}

impl TemporalitySelector for DeltaTemporalitySelector {
    fn temporality(&self, kind: InstrumentKind) -> Temporality {
        match kind {
            InstrumentKind::UpDownCounter => Temporality::Cumulative,
            InstrumentKind::ObservableUpDownCounter => Temporality::Cumulative,
            _ => Temporality::Delta,
        }
    }
}
```

**Reply by svix-jplatte** [reply]
Okay, thanks! Is that kind of `DeltaTemporalitySelector` something that you would consider including in one of the libraries (presumably the one that also defines the default one)? I'd be happy to create a PR with what you wrote plus some small adjustments and extra documentation.

Also, small tip: You can use code blocks in GitHub discussions, by writing
````md
```rust
// Rust code block with syntax highlighting
fn main() {}
```
````

**Reply by cijothomas** [reply]
> Is that kind of DeltaTemporalitySelector something that you would consider including in one of the libraries (presumably the one that also defines the default one)? I'd be happy to create a PR with what you wrote plus some small adjustments and extra documentation.

Yes. A prior attempt at that : https://github.com/open-telemetry/opentelemetry-rust/pull/1057/files which I didn't have time to finish, as we are focusing on API/SDK perf now. 

Would love to receive PR for improved docs, built-in temporality selectors, and more!
